lib/elixir/lib/string.ex

```diff
@@ -674,15 +674,17 @@ defmodule String do
   end
 
   def upcase(string, :ascii) when is_binary(string) do
-    for <<x <- string>>,
-      do: if(x >= ?a and x <= ?z, do: <<x - 32>>, else: <<x>>),
-      into: ""
+    IO.iodata_to_binary(upcase_ascii(string))
   end
 
   def upcase(string, mode) when mode in @conditional_mappings do
     String.Casing.upcase(string, [], mode)
   end
 
+  defp upcase_ascii(<<c, rest::bits>>) when c >= ?a and c <= ?z, do: [c - 32 | upcase_ascii(rest)]
```

Comment:
I think we should do

```elixir
defp upcase_ascii(<<c, rest::bits>>) when c in ?a..?z, do: ...
```

The `in` seems clearer to me and is compiled to the same thing so we should be fine performance-wise.

Reply:
Within the same minute we both commented the same

Reply:
They do not compile the same (there is an additional is_integer check which hmay be removed by the compiler or not) but we can also run into bootstrap issues, unfortunataely.

Reply:
@josevalim  so `in` creates another validation check, regardless of `..` ? I thought only integers are accepted in Range

Reply:
right, the `Kernel.".."/2` macros has the integer checks for the range

Reply:
I think the extra check is not supposed to matter in future erlang versions anyway but the biggest issue here is bootstrapping (`in` may not be available when we compile the `String` module).

Reply:
I believe the `is_integer` check on extracting data from a binary will only be eliminated on OTP 21 (I contributed that optimisation recently).